### ISSUE TITLE
2019-12-20 M1SS computation error when requesting CMAXDN introduced with 5a9f91aa0d

### ISSUE THREAD

**Issue by jhhughes256** [tensfeldt/openNCA#130]
**EDIT:** Error is caused by including "CMAXDN" in the parameterset. Please see comment on Jan 3 for description of error.

### Error

When running tc114 M1SS a computation error occurs. This error cannot be replicated with other M1SS methods. This does not occur with commit be5ef37989 and seems to be have introduced with 5a9f91aa0d. Parameterset used and error message provided below.

```r
> mct$PARAMETERLIST
[1] "AUCALL;AUCALLDN;AUCDN;AUCINFO;AUCINFOC;AUCINFODN;AUCINFOi;AUCINFP;AUCINFPC;AUCINFPDN;AUCINFPi;AUCLAST;AUCLASTi;AUCLASTC;AUCLASTCi;AUCLASTDN;AUCLASTDNi;AUCT;AUCT1_T2;AUCTAUDNi;AUCTAU;AUCTAUi;AUCTDN;AUCXPCTO;AUCXPCTOi;AUCXPCTP;AUCXPCTPi;AUMCINFO;AUMCINFOi;AUMCINFP;AUMCINFPi;AUMCLAST;AUMCLASTi;AUMCTAUi;AUMCXPTO;AUMCXPTOi;AUMCXPTP;AUMCXPTPi;CAVi;CEST;CLAST;CLASTi;CLFO;CLFOW;CLFP;CLFPW;CLFTAUi;CLFTAUWi;CMAX;CMAXi;CMAXDN;CMAXDNi;CMIN;CMINi;CMINDN;CMINDNi;CTROUGHi;CTROUGHENDi;DIi;DOSE;DOSEi;DOSEC;DOSECi;F;FREL;FRELLAST;FRELLASTi;FTAUi;KEL;KELC0;KELNOPT;KELR;KELRSQ;KELRSQA;KELTMHI;KELTMLO;LASTTIME;LASTTIMEi;MRTEVIFO;MRTEVIFOi;MRTEVIFP;MRTEVIFPi;MRTLAST;MRTLASTi;PTF;PTFi;PTR;PTRi;PTROUGHRi;PTROUGHRENDi;RACi;RACCMAXi;RACCMINi;RSSi;TAU;TAUi;THALF;THALFF;TLAG;TLAST;TLASTi;TMAX;TMAXi;TMIN;TMINi;TOLD;TOLDi;VZFO;VZFOW;VZFP;VZFPW;VZFTAUi;VZFTAUWi;FLGACCEPTPREDOSE;FLGACCEPTTMAX;FLGACCEPTKEL;FLGACCEPTTAU"

Error in value[[3L]](cond) : 
  Error in `[.data.frame`(tmp_df, , unlist(dosevar)[1]): undefined columns selected
For SDEID 42088261 
7. stop(paste0(e, "For SDEID ", unique(data_data[, map_data$SDEID])[i])) at run_M1_SS_computation.R#2884
6. value[[3L]](cond) 
5. tryCatchOne(expr, names, parentenv, handlers[[1L]]) 
4. tryCatchList(expr, classes, parentenv, handlers) 
3. tryCatch({
    if (comp_required[["DOSECi"]] || comp_required[["DOSEC"]]) {
        dose_ci <- list()
    } ... at run_M1_SS_computation.R#1050
2. run_M1_SS_computation(data = merged_data, map = map_data, method = method, 
    model_regex = model_regex, parameter_list = parameter_list, 
    return_list = return_list) at run_computation.R#2393
1. run_computation(data = d, map = mct, flag = flags, parameterset = parameterset) 
```

**Comment by jhhughes256** [comment]
This issue still occurs as of 60bb6310e9. 

The `flags` dataset for this testcase has an row with no PKDATAROWID at the bottom of the data (example below). There is no corresponding row in the concentration dataset. When this row is removed from `flags` the error still occurs.

```r
> tail(flags)
    PKDATAROWID FLGEXKEL FLGEXAUC FLGEXSDE FLGEXST PKRCOM PKPCOM FLGTIME TOLD1 TAU2 TOLD2 TAU1
425        1724        1        0        0       0               NOMINAL     0    0     0   24
426        1725        1        0        0       0               NOMINAL     0    0     0   24
427        1726        1        0        0       0               NOMINAL     0    0     0   24
428        1727        1        0        0       0               NOMINAL     0    0     0   24
429        1728        1        0        0       0               NOMINAL     0    0     0   24
430                    1        0        0       0               NOMINAL    NA   NA    NA   NA
```

**Comment by jhhughes256** [comment]
### Issue Description

This error is caused by including "CMAXDN" in the parameterset. Note that "CMAXDNi" works correctly.

It seems this is occurring due to a change in how `dosevar` is defined in the snippet starting on line 261.

#### Starting line 261

```r
  doselist <- names(parameter_indices("^DOSELIST$", names(map_data), simplify=FALSE))
  dosevar <- unlist(strsplit(map_data[,doselist], ";"))
  if(!any(duplicated(as.character(unlist(map[,dosevar]))))){
    dosevar <- map[,dosevar] 
  }
```

Originally `dosevar` was not defined conditionally, however with commit 5a9f91a the condition `if(!any(duplicated(as.character(unlist(map[,dosevar])))))` was added. I'm not certain why this change was made, as `cmax_dn` is the only function that uses this object on first inspection.

This condition prevents `dosevar` being defined according to the `mct` if the values pointed to by `DOSEi` in the `mct` are the same. While this condition by itself doesn't appear wrong, it leaves `dosevar` defined as the names of the dose items in the `mct` rather than the names of the dose columns in `tmp_df` causing a computation error when cmax_dn is called on line 1543.

**Comment by jhhughes256** [comment]
This issue has been addressed as of e681d9c. Example truncated `data_out` output shown below for tc114 M1SS.

```r
      SDEID DOSE1 DOSE2 DOSEC DOSEC1 DOSEC2 CMAX CMAX1 CMAX2 CMAXDN CMAXDN1 CMAXDN2 FLGACCEPTPREDOSE CMIN CMIN1 CMIN2 CMINDN CMINDN1
1  42088261   0.5   0.5   500    500    500 8.08  8.08  5.47  16.16   16.16   10.94                0 4.85  4.85  5.47   9.70    9.70
2 110358863   0.5   0.5   500    500    500 5.69  5.39  5.69  11.38   10.78   11.38                0 3.28  3.28  3.50   6.56    6.56
3 180819268   0.5   0.5   500    500    500 4.91  4.91  2.21   9.82    9.82    4.42                0 1.94  1.94  2.21   3.88    3.88
4 212373988   0.5   0.5   500    500    500 7.21  5.74  7.21  14.42   11.48   14.42                0 4.18  4.18  4.24   8.36    8.36
```
